optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/tri/TriConstraintStream.java

```diff
@@ -50,14 +55,147 @@
      * Exhaustively test each tuple of facts against the {@link TriPredicate}
      * and match if {@link TriPredicate#test(Object, Object, Object)} returns true.
      * <p>
-     * Important: This is slower and less scalable than {@link BiConstraintStream#join(UniConstraintStream, TriJoiner[])}
+     * Important: This is slower and less scalable than {@link BiConstraintStream#join(Class, TriJoiner[])}
      * with a proper {@link TriJoiner} predicate (such as {@link Joiners#equal(BiFunction, Function)},
      * because the latter applies hashing and/or indexing, so it doesn't create every combination just to filter it out.
      * @param predicate never null
      * @return never null
      */
     TriConstraintStream<A, B, C> filter(TriPredicate<A, B, C> predicate);
 
+    // ************************************************************************
+    // Join
+    // ************************************************************************
+
+    /**
+     * Create a new {@link QuadConstraintStream} for every combination of [A, B, C] and D.
+     * <p>
+     * Important: {@link QuadConstraintStream#filter(QuadPredicate) Filtering} this is slower and less scalable
+     * than a {@link #join(UniConstraintStream, QuadJoiner)},
+     * because it doesn't apply hashing and/or indexing on the properties,
+     * so it creates and checks every combination of [A, B] and C.
+     * @param otherStream never null
+     * @param <D> the type of the fourth matched fact
+     * @return a stream that matches every combination of [A, B, C] and D
+     */
+    default <D> QuadConstraintStream<A, B, C, D> join(UniConstraintStream<D> otherStream) {
+        return join(otherStream, new NoneQuadJoiner<>());
+    }
+
+    /**
+     * Create a new {@link QuadConstraintStream} for every combination of [A, B] and C for which the {@link QuadJoiner}
+     * is true (for the properties it extracts from all facts).
+     * <p>
+     * Important: This is faster and more scalable than a {@link #join(UniConstraintStream) join}
+     * followed by a {@link QuadConstraintStream#filter(QuadPredicate) filter},
+     * because it applies hashing and/or indexing on the properties,
+     * so it doesn't create nor checks every combination of [A, B, C] and D.
+     * @param otherStream never null
+     * @param joiner never null
+     * @param <D> the type of the fourth matched fact
+     * @return a stream that matches every combination of [A, B, C] and D for which the {@link QuadJoiner} is true
+     */
+    <D> QuadConstraintStream<A, B, C, D> join(UniConstraintStream<D> otherStream, QuadJoiner<A, B, C, D> joiner);
+
+    /**
+     * Create a new {@link QuadConstraintStream} for every combination of [A, B, C] and D.
+     * <p>
+     * Important: {@link QuadConstraintStream#filter(QuadPredicate)}  Filtering} this is slower and less scalable
```

Comment:
Double space before Filtering.

Is it consistent otherwise with the Tri/Bi/Uni versions?
Yes, this javadoc and API duplication is a pita (but we take the pain, to avoid pain on the user): as long as we have the api duplication, we can't change anything in the api or javadoc without also adjusting the other stream types consistently in the same commit.